buildengine/watch.go

```diff
@@ -60,20 +87,29 @@ func Watch(ctx context.Context, period time.Duration, moduleDirs []string, exter
 				return project.Config().Dir, project
 			})
 
+			s.mutex.Lock()
 			// Trigger events for removed projects.
-			for _, existingProject := range existingProjects {
+			for _, existingProject := range s.existingProjects {
+				if transactions, ok := s.moduleTransactions[existingProject.Project.Config().Dir]; ok && len(transactions) > 0 {
+					// Skip projects that currently have transactions
+					continue
+				}
 				existingConfig := existingProject.Project.Config()
 				if _, haveProject := projectsByDir[existingConfig.Dir]; !haveProject {
 					logger.Debugf("removed %s %q", existingProject.Project.TypeString(), existingProject.Project.Config().Key)
 					topic.Publish(WatchEventProjectRemoved{Project: existingProject.Project})
-					delete(existingProjects, existingConfig.Dir)
+					delete(s.existingProjects, existingConfig.Dir)
 				}
 			}
 
 			// Compare the projects to the existing projects.
 			for _, project := range projectsByDir {
+				if transactions, ok := s.moduleTransactions[project.Config().Dir]; ok && len(transactions) > 0 {
```

Comment:
Skipping here doesn't mean we miss out on the change. Watcher is polling on a schedule, then calculating all the hashes for all module files each time. So skipping here just means we'll notice the change the next time we poll and the transaction is over.
We probably should be listening to file changes rather than polling though (fsevents or similar)

Reply:
Ah right, makes sense. re. inotify et al, unfortunately there are no reliable cross platform libraries available for Go.